core/src/main/java/org/radargun/stages/test/Stressor.java

```diff
@@ -10,112 +11,102 @@
+ * Thread that generates the load in {@link TestStage}, created in {@link TestSetupStage}
+ * and kept in {@link RunningTest}. The thread can either execute in ramp-up state (without
+ * any statistics recorded) or in steady-state, where statistics are recorded.
+ * During execution, the thread queries {@link ConversationSelector} for the next
+ * {@link Conversation} and then executes it's {@link Conversation#run(Stressor)} method
+ * which in turn calls {@link #makeRequest(Invocation)} with certain {@link Invocation}.
  *
  * @author Radim Vansa &lt;rvansa@redhat.com&gt;
  */
 public class Stressor extends Thread {
    private static Log log = LogFactory.getLog(Stressor.class);
 
-   private final TestStage stage;
-   private final int threadIndex;
-   private final int globalThreadIndex;
-   private final OperationLogic logic;
-   private final Random random;
-   private final Completion completion;
+   private final RunningTest test;
+   private Random random;
 
-   private boolean useTransactions;
-   private int txRemainingOperations = 0;
-   private long transactionDuration = 0;
-   private Transactional.Transaction ongoingTx;
+   private ConversationSelector selector;
    private Statistics stats;
 
-   public Stressor(TestStage stage, OperationLogic logic, int globalThreadIndex, int threadIndex) {
+   private long transactionDuration = 0;
+   private boolean logTransactionExceptions = true;
+
+   public Stressor(int threadIndex, RunningTest test, boolean logTransactionExceptions) {
       super("Stressor-" + threadIndex);
-      this.stage = stage;
-      this.threadIndex = threadIndex;
-      this.globalThreadIndex = globalThreadIndex;
-      this.logic = logic;
-      this.random = ThreadLocalRandom.current();
-      this.completion = stage.getCompletion();
+      this.test = test;
+      this.logTransactionExceptions = logTransactionExceptions;
+   }
+
+   private boolean recording() {
+      return stats != null;
    }
 
    @Override
    public void run() {
+      // we cannot set this in constructor since the constructor is called in sc-main thread!
+      random = ThreadLocalRandom.current();
       try {
-         logic.init(this);
-         stats = stage.createStatistics();
-         stage.getStartLatch().await();
-
-         stats.begin();
-         runInternal();
-
-      } catch (Exception e) {
-         log.error("Unexpected error in stressor!", e);
-         stage.setTerminated();
-      } finally {
-         if (stats != null) {
-            stats.end();
-         }
-         stage.getFinishLatch().countDown();
-      }
-   }
-
-   private void runInternal() {
-      int i = 0;
-      while (completion.moreToRun(i)) {
-         Object result = null;
-         try {
-            if (useTransactions && txRemainingOperations <= 0) {
-               ongoingTx = stage.transactional.getTransaction();
-               logic.transactionStarted();
+         while (!test.isFinished()) {
```

Comment:
I'd prefer to to as little checks as possible, though, it should not incur too much overhead. These threads should get interrupted only after the finished is set to true.